src/librustc_jobserver/lib.rs

```diff
@@ -0,0 +1,292 @@
+//! rustc wants to manage its jobserver pool such that it never keeps a token
+//! around for too long if it's not being used (i.e., eagerly return tokens), so
+//! that Cargo can spawn more rustcs to go around.
+//!
+//! rustc also has a process-global implicit token when it starts, which we keep
+//! track of -- we cannot release it to Cargo, and we want to make sure that if
+//! it is released we *must* unblock a thread of execution onto it (otherwise we
+//! will deadlock on it almost for sure).
+//!
+//! So, when we start up, we have an implicit token and no acquired tokens from
+//! Cargo.
+//!
+//! We immediately on startup spawn a thread into the background to manage
+//! communication with the jobserver (otherwise it's too hard to work with the
+//! jobserver API). This is non-ideal, and it would be good to avoid, but
+//! currently that cost is pretty much required for correct functionality, as we
+//! must be able to atomically wait on both a Condvar (for other threads
+//! releasing the implicit token) and the jobserver itself. That's not possible
+//! with the jobserver API today unless we spawn up an additional thread.
+//!
+//! There are 3 primary APIs this crate exposes:
+//!  * acquire()
+//!  * release()
+//!  * acquire_from_request()
+//!  * request_token()
+//!
+//! The first two, acquire and release, are blocking functions which acquire
+//! and release a jobserver token.
+//!
+//! The latter two help manage async requesting of tokens: specifically,
+//! acquire_from_request() will block on acquiring token but will not request it
+//! from the jobserver itself, whereas the last one just requests a token (and
+//! should return pretty quickly, i.e., it does not block on some event).
+//!
+//! -------------------------------------
+//!
+//! We also have two modes to manage here. In the primary (default) mode we
+//! communicate directly with the underlying jobserver (i.e., all
+//! acquire/release requests fall through to the jobserver crate's
+//! acquire/release functions).
+//!
+//! This can be quite poor for scalability, as at least on current Linux
+//! kernels, each release on the jobserver will trigger the kernel to wake up
+//! *all* waiters instead of just one, which, if you have lots of threads
+//! waiting, is quite bad.
+//!
+//! For that reason, we have a second mode which utilizes Cargo to improve
+//! scaling here. In that mode, we have slightly altered communication with the
+//! jobserver. Instead of just blocking on the jobserver, we will instead first
+//! print to stderr a JSON message indicating that we're interested in receiving
+//! a jobserver token, and only then block on actually receiving said token. On
+//! release, we don't write into the jobserver at all, instead merely printing
+//! out that we've released a token.
```

Comment:
You know the more I read this the more I'm thinking at this point that `sccache` will be *completely* busted by this scheme... It's already bad that it doesn't share the true jobserver, but it's "good enough." After this change though I suspect we'd basically just deadlock quickly with `sccache` because rustc never actually releases tokens, and sccache doesn't know what's going on. That and there's a distrinct jobserver-per-process...

Reply:
Yes, that's true. I'm not sure how much we can do about that; one option is to work to get sccache also compatible with the new rustc protocol; potentially we can alter the protocol in some way.

For example, one thought I've had is that since in this case we know that there's a rustc on the other side, we could plausibly still have jobserver-per-rustc but place it in async read mode, where then Cargo could monitor all of them for "returned" tokens in a fairly cost-effective manner.

Is sccache working well with rustc in the new jobserver mode a hard blocker? It loosely feels like it shouldn't be, at least in the sense that we can just not use the new mode and as such be constrained to 4 threads or so when using jobserver with rustc.

Reply:
Oh I mostly brought this up as something that popped into my head, I definitely don't think this is a blocker. This is definitely something we're going to want to acknowledge for stabilization, and likely we'll want to coordinate with sccache folks before deciding what to do here.